backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/questionsubmission/QuestionSubmissionController.java

```diff
@@ -0,0 +1,105 @@
+        User user = (User) ((Authentication) principal).getPrincipal();
+
+        if (user == null)
+            throw new TutorException(AUTHENTICATION_ERROR);
+
+        return questionSubmissionService.createReview(reviewDto);
+    }
+
+    @PutMapping("/submissions/{executionId}")
+    @PreAuthorize("hasRole('ROLE_STUDENT') and hasPermission(#questionSubmissionDto.getId(),'SUBMISSION.ACCESS') and hasPermission(#executionId, 'EXECUTION.ACCESS')")
+    public QuestionSubmissionDto updateQuestionSubmission(Principal principal, @PathVariable Integer executionId, @Valid @RequestBody QuestionSubmissionDto questionSubmissionDto) {
+        User user = (User) ((Authentication) principal).getPrincipal();
+
+        if (user == null)
+            throw new TutorException(AUTHENTICATION_ERROR);
+
+        questionSubmissionDto.setUserId(user.getId());
+        return this.questionSubmissionService.updateQuestionSubmission(questionSubmissionDto.getId(), questionSubmissionDto);
+    }
+
+    @DeleteMapping("/submissions/{questionSubmissionId}")
+    @PreAuthorize("hasRole('ROLE_STUDENT') and hasPermission(#questionSubmissionId, 'SUBMISSION.ACCESS')")
+    public void removeSubmittedQuestion(@PathVariable Integer questionSubmissionId) {
+        questionSubmissionService.removeSubmittedQuestion(questionSubmissionId);
+    }
+
+    @PutMapping("/submissions/{questionSubmissionId}/topics")
+    @PreAuthorize("hasRole('ROLE_STUDENT') and hasPermission(#questionSubmissionId, 'SUBMISSION.ACCESS')")
+    public void updateQuestionSubmissionTopics(@PathVariable Integer questionSubmissionId, @RequestBody TopicDto[] topics) {
+        questionSubmissionService.updateQuestionSubmissionTopics(questionSubmissionId, topics);
+    }
+
+    @PutMapping("/submissions/{questionSubmissionId}/reviews")
+    @PreAuthorize("hasRole('ROLE_TEACHER') and hasPermission(#executionId, 'EXECUTION.ACCESS')")
+    public void toggleInReviewStatus(@PathVariable int questionSubmissionId, @Valid @RequestParam int executionId, @Valid @RequestParam boolean inReview) {
+        questionSubmissionService.toggleInReviewStatus(questionSubmissionId, inReview);
+    }
+
+    @GetMapping(value = "/submissions/{executionId}/student")
+    @PreAuthorize("hasRole('ROLE_STUDENT') and hasPermission(#executionId, 'EXECUTION.ACCESS')")
+    public List<QuestionSubmissionDto> getStudentQuestionSubmissions(Principal principal, @Valid @PathVariable int executionId) {
+        User user = (User) ((Authentication) principal).getPrincipal();
+
+        if (user == null)
+            throw new TutorException(AUTHENTICATION_ERROR);
+
+        return questionSubmissionService.getStudentQuestionSubmissions(user.getId(), executionId);
+    }
+
+    @GetMapping("/submissions/{executionId}/execution")
+    @PreAuthorize("(hasRole('ROLE_TEACHER') or hasRole('ROLE_STUDENT')) and hasPermission(#executionId, 'EXECUTION.ACCESS')")
```

Comment:
check this rule, see above

Reply:
in this case, since it's relative to course execution and not to question submission, I should check if either teacher or student are have in fact this execution, so not sure how it applies to the rule above